octopus: librbd: fixed librbd qos assert m_io_throttled failed,sometimes appeared coredump

pkulijiawei opened the pull request:
After investigation, I found that the operation of ++m_io_throttled and --m_io_throttled is done in the main thread and the timer thread respectively. Although m_io_throttled is an atomic variable, ++ The operation of m_io_throttled in the main thread may be slower than --m_io_throttled in the timer thread, so a coredump will appear when ceph_assert(m_io_throttled.load()> 0) is performed The code is as follows:
```c++
if (needs_throttle(peek_item)) {
ldout(cct, 15) << "throttling IO " << peek_item << dendl;

++m_io_throttled;
// dequeue the throttled item
ThreadPool::PointerWQ<ImageDispatchSpec<I> >::_void_dequeue();
return nullptr;
}
```
```c++
void ImageRequestWQ::handle_throttle_ready(int r, ImageDispatchSpec *item, uint64_t flag) {
CephContext *cct = m_image_ctx.cct;
ldout(cct, 15) << "r=" << r << ", " << "req=" << item << dendl;

item->set_throttled(flag);
if (item->were_all_throttled()) {
this->requeue_back(item);
ceph_assert(m_io_throttled.load() > 0);
--m_io_throttled;
this->signal();
}
}
```
Putting ceph_assert behind requeue_back can guarantee ++m_io_throttled occuring first ，This is because _void_dequeue and requeue_back need to use the same lock and _void_dequeue hold the lock.

fixed:https://tracker.ceph.com/issues/49949
Signed-off-by: pkulijiawei <lijiawei1@chinatelecom.cn>
Reviewed-by: wuxuehan@chinatelecom.cn


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

tchaikov commented:
@pkulijiawei as suggested by @smithfarm 

> Your PR title and PR description seem to really describe what the commit is doing, in detail, even including a Fixes line with a tracker link. This is really impressive and helpful, but none of this information made it into the title/description of the actual git commit, yet only the git commit becomes part of the git history.

could you add the description in the commit message, also,  could you also add a comment in the commit message mentioning that this issue only exists in octopus and nautilus?

what your commit message only includes a title of "fixed librbd qos assert m_io_throttled failed" now. could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes

pkulijiawei commented:
I am sorry i don't know what to do. I added the title as :
octopus : librbd: fixed librbd qos assert m_io_throttled failed,sometimes appeared coredump.
 and I add the description in the commit message as :

so i really don't know what to do next.


***@***.***
 
From: Kefu Chai
Date: 2021-03-24 12:41
To: ceph/ceph
CC: pkulijiawei; Mention
Subject: Re: [ceph/ceph] octopus : librbd: fixed librbd qos assert m_io_throttled failed,sometimes appeared coredump (#40354)
@pkulijiawei as suggested by @smithfarm
Your PR title and PR description seem to really describe what the commit is doing, in detail, even including a Fixes line with a tracker link. This is really impressive and helpful, but none of this information made it into the title/description of the actual git commit, yet only the git commit becomes part of the git history.
could you add the description in the commit message, also, could you also add a comment in the commit message mentioning that this issue only exists in octopus and nautilus?
what your commit message only includes a title of "fixed librbd qos assert m_io_throttled failed" now. could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

tchaikov commented:
@pkulijiawei to update the commit message, you might want to 

```
git commit --amend
```

see https://git-scm.com/docs/git-commit

and then
```
git push -f 
```
to update your remote branch, see https://git-scm.com/docs/git-push.

tchaikov commented:
@pkulijiawei  i updated your commit message

- remove the octopus prefix in the title. 
- quote the mail addresses using `<>`
- s/fixed/Fixes/
- reword the title a little bit.

dillaman commented:
@pkulijiawei Are you actually able to hit that assertion on the latest release?

pkulijiawei commented:
I find the bug at 
   octopus or nautilus. 
  
  
    
  
  
    
  
  
   
    ----------------------------
    
    
       
     李家伟
    
    中国电信云计算分公司       研发部
    
    电话：18201310782
    
    ***@***.***
    
    地址：北京市海淀区中关村新兴产业联盟大厦
    
   
   
  
    ------------------ 原始邮件 ------------------ 
    
    发件人：Jason Dillaman ***@***.***" 
    时　间：2021/03/24 20:58:41 星期三 
    收件人：cephceph ***@***.***" 
    抄送人：pkulijiawei ***@***.***", Mention ***@***.***" 
    主　题：Re: [ceph/ceph] octopus: librbd: fixed librbd qos assert m_io_throttled failed,sometimes appeared coredump (#40354) 
    
    
    @pkulijiawei Are you actually able to hit that assertion on the latest release? 
   —You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

dillaman commented:
@pkulijiawei Did you look at the linked source line? The lock is held while the `assert` is evaluated. I think you hit this bug on a much older release of Nautilus and this issue was fixed in v14.2.10 under PR #37116

pkulijiawei commented:
You are right,I saw this fixs. And I found the another way to fix this bug. I think my fix solves the problem and is simpler. 



***@***.***
 
From: Jason Dillaman
Date: 2021-03-24 21:47
To: ceph/ceph
CC: pkulijiawei; Mention
Subject: Re: [ceph/ceph] octopus: librbd: fixed librbd qos assert m_io_throttled failed,sometimes appeared coredump (#40354)
@pkulijiawei Did you look at the linked source line? The lock is held while the assert is evaluated. I think you hit this bug on a much older release of Nautilus and this issue was fixed in v14.2.10 under PR #37116
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

dillaman commented:
OK, but if there is no issues on Octopus/Nautilus right now, it's not generally a good idea to just change things for the sake of changing things. Since all this code has been re-written for Pacific, I think the correct course of action is to close this PR.

pkulijiawei commented:
okJason Dillaman ***@***.***>?2021?3?25? ??22:38???
Closed #40354.

?You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
